File: RepoTransBench/target_projects/Python/C++/maxcountryman_flask-login/public_tests/test_public_login.cpp
#include <gtest/gtest.h>
#include <string>
#include <map>

// Dummy LoginManager for illustration and API coverage
class LoginManager {
public:
    LoginManager() : login_view(nullptr), refresh_view(nullptr),
                     login_message_category("message"),
                     needs_refresh_message("Please reauthenticate to access this page."),
                     id_attribute("get_id"),
                     session_protection(nullptr) {}

    void (*anonymous_user)() = nullptr;
    const char* login_view;
    const char* refresh_view;
    std::map<std::string, std::string> blueprint_login_views;
    std::string login_message_category;
    std::string needs_refresh_message;
    std::string id_attribute;
    const char* session_protection;

    void* localize_callback = nullptr;
};

TEST(PublicLoginTest, InstanceDefaultsPublic) {
    LoginManager lm;
    // anonymous_user is a pointer; use nullptr; callable concept
    ASSERT_EQ((void*)lm.anonymous_user, (void*)nullptr);
    ASSERT_EQ(lm.login_view, nullptr);
    ASSERT_TRUE(lm.blueprint_login_views.empty());
    ASSERT_NE(lm.login_message_category.find("message"), std::string::npos);
    ASSERT_NE(lm.needs_refresh_message.find("refresh"), std::string::npos);
    ASSERT_EQ(lm.id_attribute, "get_id");
    ASSERT_TRUE((lm.session_protection == nullptr
                || std::string(lm.session_protection) == "basic"
                || std::string(lm.session_protection) == "strong"));
}

TEST(PublicLoginTest, LoginManagerCustomValuesPublic) {
    LoginManager lm;
    lm.login_view = "/custom_login";
    lm.refresh_view = "/refresh_needed";
    lm.login_message_category = "You must sign in!";
    lm.blueprint_login_views["bp2"] = "/bp2_custom_login";
    ASSERT_TRUE(std::string(lm.login_view).find("/") == 0);
    ASSERT_TRUE(std::string(lm.refresh_view).rfind("needed")+6 == std::string(lm.refresh_view).size());
    ASSERT_NE(lm.login_message_category.find("sign in"), std::string::npos);
    ASSERT_TRUE(lm.blueprint_login_views["bp2"].find("/bp2") == 0);
    // Mutate mode to strong and check
    lm.session_protection = "strong";
    ASSERT_EQ(std::string(lm.session_protection), "strong");
}

TEST(PublicLoginTest, LoginManagerAnonymousUserPublic) {
    LoginManager lm;
    void anon_func() {}
    lm.anonymous_user = anon_func;
    ASSERT_EQ((void*)lm.anonymous_user, (void*)anon_func);
}

TEST(PublicLoginTest, LocalizeCallbackPublic) {
    LoginManager lm;
    std::string called_val;
    struct F {
      static void fake_localizer(const char* txt, std::string* out) { *out = txt; }
    };
    // Simulate callback by function pointer and context
    auto cb = [](const char* txt, std::string* store) { *store = txt; };
    cb("hello", &called_val);
    ASSERT_EQ(called_val, "hello");
}

File: RepoTransBench/target_projects/Python/Rust/maxcountryman_flask-login/public_tests/test_public_login.rs
#[cfg(test)]
mod tests {
    use crate::login_manager::*;

    #[test]
    fn test_instance_defaults_public() {
        let lm = LoginManager::new();
        assert!(lm.login_view.is_none());
        assert_eq!(lm.id_attribute, "get_id".to_owned());
        assert!(lm.login_message_category.contains("message"));
        assert!(lm.needs_refresh_message.to_lowercase().contains("refresh"));
        assert!({
            lm.session_protection.is_none() ||
            lm.session_protection == Some("basic".to_owned()) ||
            lm.session_protection == Some("strong".to_owned())
        });
        assert!(lm.anonymous_user() == "AnonymousUser".to_owned());
    }

    #[test]
    fn test_login_manager_custom_values_public() {
        let mut lm = LoginManager::new();
        lm.login_view = Some("/custom_login".to_string());
        lm.refresh_view = Some("/refresh_needed".to_string());
        lm.login_message = "You must sign in!".to_string();
        lm.blueprint_login_views.insert("bp2".into(), "/bp2_custom_login".into());
        assert!(lm.login_view.as_ref().unwrap().starts_with('/'));
        assert!(lm.refresh_view.as_ref().unwrap().ends_with("needed"));
        assert!(lm.login_message.contains("sign in"));
        assert!(lm.blueprint_login_views.get("bp2").unwrap().starts_with("/bp2"));
        lm.session_protection = Some("strong".to_string());
        assert_eq!(lm.session_protection.as_deref(), Some("strong"));
    }

    #[test]
    fn test_login_manager_anonymous_user_public() {
        let mut lm = LoginManager::new();
        // Overwrite anonymous_user function
        lm.anonymous_user = std::sync::Arc::new(|| "CustomAnon".to_string());
        let res = (lm.anonymous_user)();
        assert_eq!(res, "CustomAnon");
    }

    #[test]
    fn test_localize_callback_public() {
        let mut lm = LoginManager::new();
        use std::sync::{Arc, Mutex};
        let val = Arc::new(Mutex::new(String::new()));
        let val_clone = val.clone();
        lm.localize_callback = Some(Arc::new(move |txt| {
            *val_clone.lock().unwrap() = txt.to_string();
            txt.to_string()
        }));
        if let Some(cb) = &lm.localize_callback {
            cb("hello");
        }
        assert_eq!(&*val.lock().unwrap(), "hello");
    }
}